[#49220] Adapt Project-based Boards to be rendered by Rails

aaron-contreras opened the pull request:
This PR converts the Project Boards Index Page from being rendered by Angular to being rendered by Rails and leveraging our existing `TableComponent`.

- The Angular `NewBoardModalComponent` has been removed and replaced by the Boards form created for use in the global context, making everything but the `boards#show` action be rendered by Rails.
- The `BoardsIndexPageComponent` has been removed as it's no longer in use.

**NOTES:**

- Deleting a board from within a board page's toolbar was a request performed to the API and subsequently a transition to the project's board index page by Angular. In order to have Rails handle the deletion, flash message setting and redirecting to `boards#index` (without needing to re-write this toolbar to be a form or Rails-rendered) added a JSON response that included the `redirect_url` that the component should pick up and set the page's `href`, causing the "Rails-handled redirect". If someone has a better idea on how to accomplish this before the feature freeze or if this is not a correct approach, please do tell.

- While examining the specs for Project based boards, I noticed that there was a bug with the `VersionBoardCreateService`. I've addressed it within this PR as it would pertain to project based boards as well in https://github.com/opf/openproject/commit/fb06f719e0dc88634d86e4933e58e5c45933b2d1 along with an explanation of the bug

See: https://community.openproject.org/work_packages/49220

aaron-contreras committed:
Extend coverage of `/boards` routing

Just an extension to the coverage of the routing for `/boards` ensuring
that there is proper routing to `/boards/boards#index` without a `:state`
param.

aaron-contreras committed:
Extend the `Boards::AddButtonComponent` to be usable in global and project contexts

By following the pattern of the other `AddButtonComponent`s, extending the
`#render?` and `#dynamic_path` method adds this simple ability to the ViewComponent.

aaron-contreras committed:
Convert Project Boards from being Angular-rendered to Rails-rendered

aaron-contreras committed:
Refactor: Extract common page assertion methods into base class

Since most of the expectations are shared for the boards index and
overview pages, I extracted these shared page expectations into a base
class that these page classes can inherit from and extend with their own
custom expectations/override base implementations.

aaron-contreras committed:
Add `/projects/:project_id/boards/new` route

In order to now render a boards form from a Rails context, we need
a route for it.

Adds a routing spec to assert appropriate routing and parameter matching.

aaron-contreras committed:
Add `POST /projects/:project_id/boards` route

In order to be able to post from two different contexts and have
a failed validation on the form render back the appropriate form context, we
need create actions for both global and project boards. This follows the
same pattern as in other global and project pages.

**NOTE**: A path helper naming collision occured when attempting to add
the `create` action into the `resources` macro by using `only: %i[new create]`
as both the index and create action would end up with the same path helper, being
`project_work_package_boards_path`. Given we don't really use this path helper for
`POST`s throughout the application and build it up with `url_for`, I left a comment
explaining the reasoning for it and will proceed with using `url_for` as we do elsewhere.

aaron-contreras committed:
Refactor `/boards(/*state)` into separate `index` and `show` actions

Since we're now routing with Rails and not Angular, we don't need a catch-all
`/boards*` route. We can properly route between requesting a collection and an
individual board resource.

aaron-contreras committed:
Fix: `VersionBoardCreateService` should only create queries for direct project versions

After taking a look at the specs for the previously existing `version_board_spec` which
used the functionality from the Angular modals still, it only grabs versions that are directly
pertaining to the project in hand, regardless of the `sharing` attribute on the versions.

This commit updates the `VersionBoardCreateService` to match the current/previous behavior
of the Board Creation modal in order to stay in accordance with the expected behavior.

The way it was currently working with my implementation was that it would grab all shared versions
to that project (as returned by the versions api). The Angular Version Create Service appeared
to pluck out only those which were directly belonging to the project the board was being created
for, which I missed.

aaron-contreras committed:
Convert Project Board creation modals to Rails form

Converts the modal used to create project boards into a Rails form.
This is the same form used for global create pages, but now adapted
to work both in a global and project context.

aaron-contreras committed:
Adapt project board specs to new Project Board form

Most specs were a bit tied to the modal. This commit performs a refactor
in order to work with the new board form in its place.

aaron-contreras committed:
Ensure deleting a board from the toolbar menu gets redirected to the Rails-rendered index page

aaron-contreras committed:
Add delete links to Rails Project Boards table

aaron-contreras committed:
Remove Angular `BoardsIndexPageComponent`

aaron-contreras committed:
Remove Angular `NewBoardModalComponent`

oliverguenther commented on frontend/src/app/features/boards/board/toolbar-menu/boards-toolbar-menu.directive.ts:
```diff
@@ -103,13 +103,20 @@ export class BoardsToolbarMenuDirective extends OpContextMenuTrigger {
+                `/boards/${this.board.id}`,
+                { responseType: 'json' },
+              )
+              .toPromise()
+              .then(
```
```suggestion
              .subscribe(
```

oliverguenther commented on frontend/src/app/features/boards/boards-sidebar/boards-menu.component.ts:
```diff
@@ -87,8 +86,12 @@ export class BoardsMenuComponent extends UntilDestroyedMixin implements OnInit {
-    this.opModalService.show(NewBoardModalComponent, this.injector);
+  redirectToNewBoardForm():void {
+    const newBoardFormPath = (this.currentProject && this.currentProject.identifier)
+      ? `${window.appBasePath}/projects/${this.currentProject.identifier}/boards/new`
```
I'd like to have this added to the `PathHelperService`

oliverguenther commented on modules/boards/app/controllers/boards/boards_controller.rb:
```diff
@@ -42,15 +47,35 @@ def create
+
+      respond_to do |format|
+        format.json do
+          render json: { redirect_url: project_work_package_boards_path(@project) }
```
This is fine for me 👍

oliverguenther commented on modules/boards/config/routes.rb:
```diff
@@ -3,14 +3,18 @@
+    # Adding the `create` action into the above `resources` macro would
+    # result in a name collision between it and `get /boards(/*state)`
+    # as it would result in both being named `project_work_package_boards`
+    post '/boards', to: 'boards/boards#create'
```
I'm not sure I follow, isn't the naming expected? You did remove the `boards/(*state)` route after all, so `show` now also has the same route helper?

aaron-contreras replied:
It's a stale comment. The issue would arise before I split `boards/(*state)` out into individual `index` and `show` actions. This "special case route" can even be removed an added in into the `resources` macro :)

oliverguenther reviewed (changes requested):
Behavior looking good to me, there are a few minor remarks.

With this PR, we can remove parts of the angular based action services (for creating new boards at least). I can take a look into that separately.

oliverguenther committed:
Replace toPromise().then with subscribe

oliverguenther committed:
Replace path with existing path in PathHelperService

oliverguenther committed:
Remove dead code now that creation is handled in backend

aaron-contreras committed:
Replace special `POST /boards` route with `resources` macro

The special case for the `POST /boards` route was only needed due
to the collisions that would happen when sharing an index and show
route with the same path matcher. Once these routes were split, there
was no longer a path helper naming collission so we can ammend the
`resources` macro with `only: %i[create]` now as well, making better
use of it and cleaning up the routes.
